miasm2/expression/simplifications_common.py

```diff
@@ -1132,31 +1245,143 @@ def simp_ext_cst(expr_s, expr):
+        # C.zeroExt(X)[A:B] => C[A:B] if B <= size(C)
+        return arg[expr.start:expr.stop]
+    if expr.start == 0:
+        # A.zeroExt(X)[0:Y] => A.zeroExt(Y)
+        return arg.zeroExtend(expr.stop)
+    return expr
+
+def simp_slice_of_op_ext(expr_s, expr):
+    """(X.zeroExt() + ... + Int)[0:8] => X + ... + int[:]"""
+    if expr.start != 0:
+        return expr
+    src = expr.arg
+    if not src.is_op("+"):
+        return expr
+    is_ok = True
+    for arg in src.args:
+        if arg.is_int():
+            continue
+        if (arg.is_op() and
+            arg.op.startswith("zeroExt") and
+            arg.args[0].size == expr.stop):
+            continue
+        is_ok = False
+        break
+    if not is_ok:
+        return expr
+    args = [expr_s(arg[:expr.stop]) for arg in src.args]
+    return ExprOp("+", *args)
+
+
+def simp_cond_logic_ext(expr_s, expr):
+    """(X.zeroExt() + ... + Int) ? A:B => X + ... + int[:] ? A:B"""
+    cond = expr.cond
+    if not cond.is_op():
+        return expr
+    if cond.op not in ["&", "^", "|"]:
         return expr
-    return arg
+    is_ok = True
+    sizes = set()
+    for arg in cond.args:
+        if arg.is_int():
+            continue
+        if (arg.is_op() and
+            arg.op.startswith("zeroExt")):
+            sizes.add(arg.args[0].size)
+            continue
+        is_ok = False
+        break
+    if not is_ok:
+        return expr
+    if len(sizes) != 1:
+        return expr
+    size = list(sizes)[0]
+    args = [expr_s(arg[:size]) for arg in cond.args]
+    cond = ExprOp(cond.op, *args)
+    return ExprCond(cond, expr.src1, expr.src2)
+
+
+def simp_cond_sign_bit(_, expr):
+    """(a & .. & 0x80000000) ? A:B => (a & ...) <s ? A:B"""
+    cond = expr.cond
+    if not cond.is_op('&'):
+        return expr
+    last = cond.args[-1]
+    if not last.is_int(1 << (last.size - 1)):
+        return expr
+    zero = ExprInt(0, expr.cond.size)
+    if len(cond.args) == 2:
+        args = [cond.args[0], zero]
+    else:
+        args = [ExprOp('&', *list(cond.args[:-1])), zero]
+    cond = ExprOp(TOK_INF_SIGNED, *args)
+    return ExprCond(cond, expr.src1, expr.src2)
+
 
-def simp_add_multiple(expr_s, expr):
-    # X + X => 2 * X
-    # X + X * int1 => X * (1 + int1)
-    # X * int1 + (- X) => X * (int1 - 1)
-    # X + (X << int1) => X * (1 + 2 ** int1)
-    # Correct even if addition overflow/underflow
+def simp_test_signext_inf(expr_s, expr):
+    """A.signExt() <s int => A <s int[:]"""
+    if not (expr.is_op(TOK_INF_SIGNED) or expr.is_op(TOK_INF_EQUAL_SIGNED)):
+        return expr
+    arg, cst = expr.args
+    if not (arg.is_op() and arg.op.startswith("signExt")):
+        return expr
+    if not cst.is_int():
+        return expr
+    base = arg.args[0]
+    tmp = int(mod_size2int[cst.size](int(cst)))
+    if -(1 << (base.size - 1)) <= tmp < (1 << (base.size - 1)):
+        # Can trunc integer
+        return ExprOp(expr.op, base, expr_s(cst[:base.size]))
```

Comment:
I'm not sure about this one, but as `tmp` comes from a bit to Python int translation, it should probably have to go from a Python int back to a bit representation at the end (using `mod_size2int` like transformation). I'm not sure the truncation is equivalent (it might be giving the range check).
You probably had a different view of this in mind, which is more explanatory of why it is working. If you can share it, that would help me to be sure it is indeed working :)

Reply:
Hum, I cannot parse:
```
but as tmp comes from a bit to Python int translation
```

The rule is this one:
```
A.signExt() <s int => A <s int[:]
```
Example:
```
ExprInt(-3, 8).signExt(32) <s ExprInt(12, 32)
```
Can be reduced in:
```
ExprInt(-3, 8) <s ExprInt(12, 8)
```
Idem for a negative constant:
```
ExprInt(-12, 8).signExt(32) <s ExprInt(-3, 32)
```
Can be reduced in:
```
ExprInt(-12, 8) <s ExprInt(-3, 8)
```
To check this, we need to verify if the constant (in signed form) fit in it's future truncated signed range.

If the constant fits in its future truncated signed range, this means the signed extend of this constant always gives the original constant value.
Exemple:
```
ExprId("A", 8).signExt(32) <s ExprInt(-3, 32)
```
* A is in range `[-128, 127]`
* `ExprInt(-3, 32)` is in this range `[-128, 127]`
* so we can compare against `ExprInt(-3, 8)`
* And effectively, `ExprInt(-3, 8).signExt(32)` gives `ExprInt(-3, 32)`

Reply:
Note: we could better:
* If the value is bigger than the original range, the case is *always* true.
* if the value is lesser than the original range, the case is *always* false

Everything is the same for `simp_test_zeroext_inf`
Maybe this one os simpler as in unsigned domain.

These previous rules may be ok for unsigned as well:
```
ExprInt('a', 8).zeroExt(32) <u ExprInt(300, 32)
```
is always true.

Reply:
I corrected a tipo in `simp_test_zeroext_inf` for size2uint

Reply:
Alright, I agree with you. I was missing that if the cst can fit in the range, it means it can be rewritten as small_cst.signExtend() (and then can be truncated).
Thanks for the rewording.